krm-functions/nfdeploy-fn/common/mutator.go

```diff
@@ -0,0 +1,206 @@
+		return false, err
+	}
+
+	return nfDeployFn.sdk.Run()
+}
+
+func (h *NfDeployFn[T1]) ClusterContextCallBackFn(o *fn.KubeObject) error {
+	var cluster infrav1alpha1.ClusterContext
+	err := o.As(&cluster)
+	if err != nil {
+		return err
+	}
+	h.site = *cluster.Spec.SiteCode
+	return nil
+}
+
+func (h *NfDeployFn[T1]) CapacityContextCallBackFn(o *fn.KubeObject) error {
+	var capacity nephioreqv1alpha1.Capacity
+	err := o.As(&capacity)
+	if err != nil {
+		return err
+	}
+
+	h.capacity = capacity
+	return nil
+}
+
+func (h *NfDeployFn[T1]) InterfaceCallBackFn(o *fn.KubeObject) error {
+	var itfce *nephioreqv1alpha1.Interface
+	err := o.As(&itfce)
+	if err != nil {
+		return err
+	}
+
+	itfcIPAllocStatus := itfce.Status.IPAllocationStatus
+	itfcVlanAllocStatus := itfce.Status.VLANAllocationStatus
+
+	if itfcIPAllocStatus == nil || itfcVlanAllocStatus == nil {
+		return nil
+	}
+
+	itfcConfig := nephiodeployv1alpha1.InterfaceConfig{
+		Name: itfce.Name,
+		IPv4: &nephiodeployv1alpha1.IPv4{
+			Address: itfcIPAllocStatus.AllocatedPrefix,
+			Gateway: &itfcIPAllocStatus.Gateway,
+		},
+		VLANID: &itfcVlanAllocStatus.AllocatedVlanID,
+	}
+
+	h.SetInterfaceConfig(itfcConfig, itfce.Spec.NetworkInstance.Name)
+	return nil
+}
+
+func (h *NfDeployFn[T1]) DnnCallBackFn(o *fn.KubeObject) error {
+	var dnnReq nephioreqv1alpha1.DataNetwork
+	err := o.As(&dnnReq)
+	if err != nil {
+		return err
+	}
+
+	if dnnReq.Status.Pools == nil {
+		return nil
+	}
+
+	var pools []nephiodeployv1alpha1.Pool
+	for _, pool := range dnnReq.Status.Pools {
+		pools = append(pools, nephiodeployv1alpha1.Pool{Prefix: pool.IPAllocation.AllocatedPrefix})
+	}
+
+	dnn := nephiodeployv1alpha1.DataNetwork{
+		Name: &dnnReq.Spec.NetworkInstance.Name,
+		Pool: pools,
+	}
+
+	h.AddDNNToNetworkInstance(dnn, dnnReq.Spec.NetworkInstance.Name)
+
+	return nil
+}
+
+func (h *NfDeployFn[T1]) GenerateResourceFn(nfDeploymentObj *fn.KubeObject, _ fn.KubeObjects) (*fn.KubeObject, error) {
+	var err error
+
+	if nfDeploymentObj == nil {
+		nfDeploymentObj = fn.NewEmptyKubeObject()
+
+		err = nfDeploymentObj.SetAPIVersion(nephiodeployv1alpha1.GroupVersion.String())
+		if err != nil {
+			return nil, err
+		}
+
+		err = nfDeploymentObj.SetKind(h.gvk.Kind)
+		if err != nil {
+			return nil, err
+		}
+
+		err = nfDeploymentObj.SetName(h.pkgName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	nfKoExt, err := kubeobject.NewFromKubeObject[T1](nfDeploymentObj)
+	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	nfSpec := &nephiodeployv1alpha1.NFDeploymentSpec{}
+
+	h.FillCapacityDetails(nfSpec)
+
+	if err != nil {
+		return nil, err
+	}
+
+	for networkInstanceName, itfceConfig := range h.interfaceConfigsMap {
+		h.AddInterfaceToNetworkInstance(itfceConfig.Name, networkInstanceName)
+	}
+
+	nfSpec.Interfaces = h.GetAllInterfaceConfig()
+	nfSpec.NetworkInstances = h.GetAllNetworkInstance()
+
+	err = nfKoExt.SetSpec(nfSpec)
+
+	nf, _ := nfKoExt.GetGoStruct()
+	fmt.Println(nf)
```

Comment:
For debugging? May be good to check the error code before printing?

Reply:
Sorry, should have removed it. Thanks.